include/pybind11/async.h

```diff
@@ -0,0 +1,208 @@
+        }
+
+        /// Construct a async_function from a class method (const, no ref-qualifier)
+        template <typename Return, typename Class, typename... Arg, typename... Extra>
+        async_function(Return (Class::*f)(Arg...) const, const Extra&... extra) {
+            initialize([f](const Class *c, Arg... args) -> Return { return (c->*f)(std::forward<Arg>(args)...); },
+                    (Return (*)(const Class *, Arg ...)) nullptr, extra...);
+        }
+
+        /// Construct a async_function from a class method (const, lvalue ref-qualifier)
+        /// A copy of the overload for const functions without explicit ref-qualifier
+        /// but with an added `&`.
+        template <typename Return, typename Class, typename... Arg, typename... Extra>
+        async_function(Return (Class::*f)(Arg...) const&, const Extra&... extra) {
+            initialize([f](const Class *c, Arg... args) -> Return { return (c->*f)(args...); },
+                    (Return (*)(const Class *, Arg ...)) nullptr, extra...);
+        }
+
+    protected:
+        template <typename Func, typename Return, typename... Args, typename... Extra>
+        void initialize(Func &&f, Return (*)(Args...), const Extra&... extra) {
+            // create a new lambda which spawns an async thread running the original function
+            auto proxy = [f](Args... args) -> Awaitable* {
+                auto thread_func = [f](Args... args) {
+                    auto result = f(std::forward<Args>(args) ...);
+
+                    py::gil_scoped_acquire gil;
+
+                    auto py_result = py::cast(result);
+                    return py_result;
+                };
+                auto bound_thread_func = std::bind(thread_func, std::forward<Args>(args)...);
+
+                auto future = std::async(std::launch::async, bound_thread_func);
+                auto awaitable = new Awaitable(future);
+
+                return awaitable;
+            };
+
+            // initialize using the new lambda function
+            cpp_function::initialize(
+                std::forward<decltype(proxy)>(proxy),
+                (detail::function_signature_t<decltype(proxy)> *) nullptr,
+                extra...
+                );
+        }
+
+        template <typename Func, typename... Args, typename... Extra>
+        void initialize(Func &&f, void (*)(Args...), const Extra&... extra) {
+            // create a new lambda which spawns an async thread running the original function
+            auto proxy = [f](Args... args) -> Awaitable* {
+                auto thread_func = [f](Args... args) {
+                    f(std::forward<Args>(args) ...);
+
+                    py::gil_scoped_acquire gil;
+
+                    auto py_result = py::cast(Py_None);
+                    return py_result;
+                };
+                auto bound_thread_func = std::bind(thread_func, std::forward<Args>(args)...);
```

Comment:
std::bind is now forbidden in the code base (merge in the new clang-tidy rules). Should be a lambda instead.

Reply:
fixed, thanks!

Reply:
NOT!

i tried the two following but it's causing segfaults:

``` c++
auto bound_thread_func = [&args..., &thread_func]{
    return thread_func(std::forward<Args>(args)...);
};
```

``` c++
auto bound_thread_func = [&args..., &thread_func]{
    return thread_func(args...);
};
```

i added a minimal reproducer down below if you have an idea..

Reply:
Clang-Tidy can automatically rewrite this function for you.
```
auto bound_thread_func = [&thread_func](auto&& args){
     return thread_func(std::forward<Args>(args)...);
};